examples/api/lib/material/input_chip/input_chip.1.dart

```diff
@@ -0,0 +1,434 @@
+                },
+              ),
+            ),
+        ],
+      ),
+    );
+  }
+
+  Future<void> _onSearchChanged(String value) async {
+    final List<String> results = await _suggestionCallback(value);
+    setState(() {
+      _suggestions = results
+          .where((String topping) => !_toppings.contains(topping))
+          .toList();
+    });
+  }
+
+  Widget _chipBuilder(BuildContext context, String topping) {
+    return InputChip(
+      key: ObjectKey(topping),
+      label: Text(topping),
+      avatar: CircleAvatar(
+        child: Text(topping[0].toUpperCase()),
+      ),
+      onDeleted: () => _deleteChip(topping),
+      onSelected: (_) => _onChipTapped(topping),
+      materialTapTargetSize: MaterialTapTargetSize.shrinkWrap,
+      padding: const EdgeInsets.all(2),
+    );
+  }
+
+  Widget _suggestionBuilder(BuildContext context, String topping) {
+    return ListTile(
+      key: ObjectKey(topping),
+      leading: CircleAvatar(
+        child: Text(topping[0].toUpperCase()),
+      ),
+      title: Text(topping),
+      onTap: () => _selectSuggestion(topping),
+    );
+  }
+
+  void _selectSuggestion(String topping) {
+    setState(() {
+      _toppings.add(topping);
+      _suggestions = <String>[];
+    });
+  }
+
+  void _onChipTapped(String topping) {}
+
+  void _deleteChip(String topping) {
+    setState(() {
+      _toppings.remove(topping);
+      _suggestions = <String>[];
+    });
+  }
+
+  void _onSubmitted(String text) {
+    if (text.trim().isNotEmpty) {
+      setState(() {
+        _toppings = <String>{..._toppings, text.trim()};
+      });
+    } else {
+      _chipFocusNode.unfocus();
+      setState(() {
+        _toppings = <String>{};
+      });
+    }
+  }
+
+  void _onChanged(Set<String> data) {
+    setState(() {
+      _toppings = data;
+    });
+  }
+
+  FutureOr<List<String>> _suggestionCallback(String text) {
+    if (text.isNotEmpty) {
+      return _pizzaToppings.where((String topping) {
+        return topping.toLowerCase().contains(text.toLowerCase());
+      }).toList();
+    }
+    return const <String>[];
+  }
+}
+
+class ChipsInput<T> extends StatefulWidget {
+  const ChipsInput({
+    super.key,
+    required this.values,
+    this.decoration = const InputDecoration(),
+    required this.chipBuilder,
+    required this.onChanged,
+    this.onChipTapped,
+    this.onSubmitted,
+    this.onTextChanged,
+    this.focusNode,
+  });
+
+  final Set<T> values;
+  final InputDecoration decoration;
+  final FocusNode? focusNode;
+
+  final ValueChanged<Set<T>> onChanged;
+  final ValueChanged<T>? onChipTapped;
+  final ValueChanged<String>? onSubmitted;
+  final ValueChanged<String>? onTextChanged;
+
+  final Widget Function(BuildContext context, T data) chipBuilder;
+
+  @override
+  ChipsInputState<T> createState() => ChipsInputState<T>();
+}
+
+class ChipsInputState<T> extends State<ChipsInput<T>>
+    implements TextInputClient {
```

Comment:
@justinmc - is there a simpler way to do this than implementing TextInputClient?

Reply:
Yeah we shouldn't be recommending people do this in our examples, because this will result in the loss of lots of functionality in EditableText.  For example, the delete key doesn't work in this example with a hardware keyboard.

I think the best way to do this would be to extend TextEditingController and override buildTextSpan and then use a TextField like normal.  You can insert your chips inline into the text as WidgetSpans in buildTextSpan.

CC @Renzo-Olivares who is the real expert in this.